ruby/BUILD.bazel

```diff
@@ -3,30 +3,180 @@
+    extension = "lib/google/protobuf_c.so",
+    deps = glob(["ext/google/protobuf_c/*"]),
+    target_compatible_with = select({
+        ":java_ruby": ["@platforms//:incompatible"],
+        "@platforms//os:osx": ["@platforms//:incompatible"],
+        "//conditions:default": [],
+    }),
+)
+
+internal_ruby_extension(
+    name = "protobuf_java",
+    extension = "lib/google/protobuf_java.jar",
+    deps = glob(["src/**/*.java"]),
+    target_compatible_with = select({
+        ":java_ruby": [],
+        "//conditions:default": ["@platforms//:incompatible"],
+    }),
+)
+
+filegroup(
+    name = "protobuf",
+    srcs = [
+        ":srcs",
+        "//third_party/utf8_range:all_files",
+    ] + select({
+        ":java_ruby": [":protobuf_java"],
+        "@bazel_tools//src/conditions:darwin": [":protobuf_c_mac"],
+        "//conditions:default": [":protobuf_c"],
+    }),
+    visibility = [
+        "//conformance:__subpackages__",
+        "//ruby:__subpackages__",
+    ],
+)
+
+################################################################################
+# Tests
+################################################################################
+
+filegroup(
+  name = "tests",
+  srcs = glob(["tests/*.rb"]),
+)
+
+filegroup(
+  name = "test_protos",
+  srcs = glob(["tests/*.proto"]),
+)
+
+internal_ruby_proto_library(
+    name = "test_ruby_protos",
+    srcs = [":test_protos"],
+    proto_deps = ["//:well_known_protos"],
+    includes = [".", "src", "ruby/tests"],
+)
+
+inline_sh_test(
+    name = "test",
+    srcs = [
+        "Rakefile",
+    ],
+    deps = [
+        ":protobuf",
+        ":test_ruby_protos",
+        ":tests",
+        "//:well_known_ruby_protos",
+    ],
+    cmd = """
+        pushd `dirname $(location Rakefile)`
+        RUBYLIB=../src:tests:. BAZEL=true rake test
+        popd
+    """,
+)
+
+inline_sh_test(
+    name = "gc_test",
+    srcs = [
+        "Rakefile",
+    ],
+    deps = [
+        ":protobuf",
+        ":test_ruby_protos",
+        ":tests",
+        "//:well_known_ruby_protos",
+    ],
+    cmd = """
+        pushd `dirname $(location Rakefile)`
+        RUBYLIB=../src:tests:. BAZEL=true rake gc_test
+        popd
+    """,
+)
+
+conformance_test(
+    name = "conformance_test",
+    failure_list = "//conformance:failure_list_ruby.txt",
+    testee = "//conformance:conformance_ruby",
+    text_format_failure_list = "//conformance:text_format_failure_list_ruby.txt",
+    target_compatible_with = select({
+        ":java_ruby": ["@platforms//:incompatible"],
+        "//conditions:default": [],
+    }),
+)
+
+conformance_test(
+    name = "conformance_test_jruby",
+    failure_list = "//conformance:failure_list_jruby.txt",
+    testee = "//conformance:conformance_ruby",
+    text_format_failure_list = "//conformance:text_format_failure_list_jruby.txt",
+    target_compatible_with = select({
+        ":java_ruby": [],
+        "//conditions:default": ["@platforms//:incompatible"],
+    }),
+)
```

Comment:
How can I run `bazel test --cache_test_results=no //ruby:conformance_test_jruby` locally now? It gives:

```
WARNING: /Users/chenzili/Brittani/protobuf/ruby/BUILD.bazel:99:28: in _proto_gen rule //ruby:test_ruby_protos_deps_genproto: target '//ruby:test_ruby_protos_deps_genproto' depends on deprecated target '//:well_known_protos': Prefer :well_known_type_protos instead.
ERROR: Target //ruby:conformance_test_jruby is incompatible and cannot be built, but was explicitly requested.
Dependency chain:
    //ruby:conformance_test_jruby   <-- target platform (@local_config_platform//:host) didn't satisfy constraint @platforms//:incompatible
INFO: Elapsed time: 0.115s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (1 packages loaded, 69 targets configured)
FAILED: Build did NOT complete successfully (1 packages loaded, 69 targets configured)
```

Reply:
cc @mkruskal-google @deannagarcia

Reply:
Ruby is in an awkward state right now, but I'm currently working on cleaning it up (https://github.com/protocolbuffers/protobuf/pull/10525).  Until then, this command should allow you to run the jruby conformance tests:
`bazel test //ruby:conformance_test_jruby --define=ruby_platform=java --action_env=PATH --action_env=GEM_PATH --action_env=GEM_HOME`

After my PR this should be as simple as running `bazel test //ruby/...` on a system where jruby is currently installed

Reply:
The main issues here are that:
1) Bazel can't currently determine if it should run the jruby or ruby conformance tests
2) Ruby depends on some environment variables that Bazel strips by default

Reply:
[out.txt](https://github.com/protocolbuffers/protobuf/files/9570768/out.txt)

Interesting. Here is the output when I was debugging #10454. I ran:

```
bazel test //ruby:conformance_test_jruby --define=ruby_platform=java --action_env=PATH --action_env=GEM_PATH --action_env=GEM_HOME
```

... and it seems that still c ext is executed:

```
Starting local Bazel server and connecting to it...
WARNING: /Users/chenzili/Brittani/protobuf/ruby/BUILD.bazel:99:28: in _proto_gen rule //ruby:test_ruby_protos_deps_genproto: target '//ruby:test_ruby_protos_deps_genproto' depends on deprecated target '//:well_known_protos': Prefer :well_known_type_protos instead.
INFO: Analyzed target //ruby:conformance_test_jruby (102 packages loaded, 2180 targets configured).
INFO: Found 1 test target...
ERROR: /Users/chenzili/Brittani/protobuf/ruby/BUILD.bazel:59:24: Executing genrule //ruby:protobuf_java failed: (Exit 1): bash failed: error executing command /bin/bash -c ... (remaining 1 argument skipped)

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
rm -f 
rake aborted!
Don't know how to build task 'ext/google/protobuf_c/extconf.rb' (See the list of available tasks with `rake --tasks`)

Tasks: TOP => default => build => compile => compile:arm64-darwin21 => compile:protobuf_c:arm64-darwin21 => copy:protobuf_c:arm64-darwin21:3.1.2 => tmp/arm64-darwin21/protobuf_c/3.1.2/protobuf_c.bundle => tmp/arm64-darwin21/protobuf_c/3.1.2/Makefile
(See full trace by running task with --trace)
/private/var/tmp/_bazel_chenzili/56fb92eeedb5ac2159dafd3ab4543886/sandbox/darwin-sandbox/1/execroot/com_google_protobuf/ruby /private/var/tmp/_bazel_chenzili/56fb92eeedb5ac2159dafd3ab4543886/sandbox/darwin-sandbox/1/execroot/com_google_protobuf
mkdir -p tmp/arm64-darwin21/protobuf_c/3.1.2
Target //ruby:conformance_test_jruby failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 3.985s, Critical Path: 0.41s
INFO: 4 processes: 4 internal.
FAILED: Build did NOT complete successfully
//ruby:conformance_test_jruby                                   FAILED TO BUILD

FAILED: Build did NOT complete successfully
```

.. so I hard code `RUBY_PLATFORM = "java"` in `ruby/Rakefile` and rerun, then I got the output that logics in c ext "Error occurred during parsing" is printed.

Reply:
What shows up if you run `ruby --version`?

Reply:
```
$ ruby --version
jruby 9.3.7.0 (2.6.8) 2022-08-16 c79ef237e0 OpenJDK 64-Bit Server VM 17.0.3+7 on 17.0.3+7 +jit [arm64-darwin]
```

I use `rbenv` to manage the version and run `rbenv local jruby-9.3.7.0` at the root path of `protobuf`.

Reply:
Yes. `rbenv global` works. Thank you!

It will be helpful if you can give a review to #10454 :)